src/runtime/hexagon/sim/hexagon_device_sim.cc

```diff
@@ -0,0 +1,1419 @@
+    src++;
+    len--;
+  }
+  CHECK_EQ(len, 0);
+}
+
+void HexagonSimulator::SendMsg(Message& m, const void* data, bool show_dbg) {
+  auto run = [this](bool report_cycles) {
+    HEXAPI_CoreState core = HEX_CORE_RESET;
+    HEX_4u_t result;
+    HEX_8u_t cycles0, cycles1;
+    if (report_cycles) {
+      CHECK_EQ(sim_->GetSimulatedCycleCount(&cycles0), HEX_STAT_SUCCESS);
+    }
+
+    core = sim_->Run(&result);
+    CHECK_EQ(core, HEX_CORE_BREAKPOINT);
+    if (report_cycles) {
+      CHECK_EQ(sim_->GetSimulatedCycleCount(&cycles1), HEX_STAT_SUCCESS);
+      LOG(INFO) << "host: execution took " << (cycles1 - cycles0) << " cycles";
+    }
+  };
+
+  // Send the message request.
+  Message r = {kMsgReq, m.len, 0u};
+  CopyToV(message_buffer_v_, &r, sizeof(r));
+  run(false);
+
+  // Receive the acknowledgement with the address for the payload.
+  CopyFromV(&r, message_buffer_v_, sizeof(r));
+  CHECK_EQ(r.code, kMsgAck);
+  CHECK_GE(r.len, m.len);
+
+  // Send the actual message.
+  m.va = r.va;
+  CopyToV(message_buffer_v_, &m, sizeof(m));
+  if (m.len > 0) CopyToV(r.va, data, m.len);
+  run(show_dbg);
+
+  // Receive the return data.
+  CopyFromV(&m, message_buffer_v_, sizeof(m));
+  CHECK_EQ(m.code, kNone);
+}
+
+HexagonSimulator::HexagonSimulator(bool enable_queuing) {
+  task_queuing_ = enable_queuing;
+
+  // The simulator argument string is in the form:
+  //   <cpu_ver> <optional_arguments>
+  // The optional arguments are seperated with spaces:
+  // Ex: --hvx_length 128 --memfill 0 --timing -m output.txt
+  const char* sim_args_env = std::getenv("HEXAGON_SIM_ARGS");
+  if (sim_args_env == nullptr) sim_args_env = "";
+  auto sim_args_iss = std::istringstream(std::string(sim_args_env));
+  using iterator = std::istream_iterator<std::string>;
+  auto sim_args = string_list(iterator(sim_args_iss), iterator());
+
+  std::string target_str =
+      !sim_args.empty() ? *detail::pop_front(sim_args) : std::string("v66");
+
+  arch_ = target_str;
+  sim_ =
+      detail::make_unique<HexagonWrapper>(detail::non_const_str(target_str));
+  LOG(INFO) << "HexagonSimulator: Core version: " << arch_;
+
+  // Locate the sim_dev binary in PATH, or in the current working directory.
+  llvm::StringRef sim_dev = "sim_dev";
+  detail::MaybeString path_sim_dev =
+      llvm::sys::Process::FindInEnvPath("PATH", sim_dev);
+  if (!path_sim_dev) {
+    if (!llvm::sys::fs::exists(sim_dev)) {
+      LOG(FATAL) << "Cannot find sim_dev in PATH.";
+    }
+    path_sim_dev = sim_dev.str();
+  }
+
+  CHECKED_CALL(ConfigureExecutableBinary, path_sim_dev->c_str());
+
+  std::vector<std::string> app_args = {*path_sim_dev};
+  if (char* ev = getenv("ADSP_LIBRARY_PATH")) {
+    app_args.push_back("-L");
+    app_args.push_back(ev);
+  }
+  sim_dev_args_ = detail::non_const_str(app_args);
+  CHECKED_CALL(ConfigureAppCommandLine, sim_dev_args_.size(), sim_dev_args_);
+
+  LOG(INFO) << "HexagonSimulator: ConfigureHVXLength: 128";
+  CHECKED_CALL(ConfigureHVXLength, 128);
```

Comment:
I am concerning here. When we create Hexagon target and use HVX 64, but we write hard code of 128 here. We can not find the problem easily.  I didn't think of good solution before too. If we don't have a better way, could we add comment in `def hexagon(cpu_ver='v66', sim_args=None, hvx=128):` to prompt users if you change the hvx length, we should change the simulator here?

Reply:
This was actually only setting the default HVX length.  If `--hvx_length` was specified in `sim_args, it would be processed in `Configure` below, resetting the HVX length to the new value.

I changed the processing of `sim_args` to always add `--hvx_length` that matches `hvx` (it already printed a warning if both were given and didn't match).  Now, configuring the default HVX length here is not needed, so I removed it.